pkg/queue/sharedmain/handlers.go

```diff
@@ -139,3 +144,11 @@ func withFullDuplex(h http.Handler, enableFullDuplex bool, logger *zap.SugaredLo
 		h.ServeHTTP(w, r)
 	})
 }
+
+func withRequestCounter(h http.Handler, pendingRequests *atomic.Int32) http.Handler {
```

Comment:
Can we create an struct that implements http.Handler and it holds the atomic counter - that would make this reusable

Your handler can hold the next handler so you can call `h.ServeHTTP(w, r)`